// sigstore/dist/bundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildDSSEBundle = exports.buildBlobBundle = exports.ATTESTATION_TYPE_DSSE = exports.ATTESTATION_TYPE_BLOB = void 0;
exports.ATTESTATION_TYPE_BLOB = 'attestation/blob';
exports.ATTESTATION_TYPE_DSSE = 'attestation/dsse';
const buildBlobBundle = (digest, signature, certificate, rekorEntry) => ({
    attestationType: exports.ATTESTATION_TYPE_BLOB,
    attestation: {
        payloadHash: digest,
        payloadHashAlgorithm: 'sha256',
        signature: signature,
    },
    certificate: certificate,
    signedEntryTimestamp: rekorEntry.verification.signedEntryTimestamp,
    integratedTime: rekorEntry.integratedTime,
    logID: rekorEntry.logID,
    logIndex: rekorEntry.logIndex,
});
exports.buildBlobBundle = buildBlobBundle;
const buildDSSEBundle = (envelope, certificate, rekorEntry) => ({
    attestationType: exports.ATTESTATION_TYPE_DSSE,
    attestation: envelope,
    certificate: certificate,
    signedEntryTimestamp: rekorEntry.verification.signedEntryTimestamp,
    integratedTime: rekorEntry.integratedTime,
    logID: rekorEntry.logID,
    logIndex: rekorEntry.logIndex,
});
exports.buildDSSEBundle = buildDSSEBundle;
